frb_dart/lib/src/ffi/io.dart

```diff
@@ -51,26 +37,137 @@ class DartApiDl {
-  @ffi.Int32()
-  external int len;
+  @ffi.Bool()
+  external bool isSuccess;
 
-  /// Not to be used by normal users, but has to be public for generated code
-  @ffi.Uint8()
-  external int success;
+  external ffi.Pointer<ffi.Void> ptr;
+
+  dynamic intoDart() => data.intoDart();
+}
 
-  Uint8List? get buffer {
-    if (ptr.address == 0) {
-      return null;
+extension DartCObjectWireSyncReturn on DartCObject {
+  dynamic intoDart() {
+    switch (ty) {
+      case DartCObjectType.DartNull:
+        return null;
+      case DartCObjectType.DartBool:
+        return value.as_bool;
+      case DartCObjectType.DartInt32:
+        return value.as_int32;
+      case DartCObjectType.DartInt64:
+        return value.as_int64;
+      case DartCObjectType.DartDouble:
+        return value.as_double;
+
+      case DartCObjectType.DartString:
+        // `DartCObject` strings being encoded with std::ffi::CString assert us nul-termination.
+        // See [allo-isolate's String::into_dart](https://github.com/sunshine-protocol/allo-isolate/blob/71b9760993d64ef46794176ca276d1cc637b2599/src/into_dart.rs#L106)
+        // and [std::ffi::CString](https://doc.rust-lang.org/nightly/std/ffi/struct.CString.html)
+        int len = 0;
+        while (value.as_string.elementAt(len).value != 0) {
+          len++;
+        }
+        return utf8.decode(value.as_string.cast<ffi.Uint8>().asTypedList(len));
+
+      case DartCObjectType.DartArray:
+        return List.generate(value.as_array.length,
+            (i) => value.as_array.values.elementAt(i).value.ref.intoDart());
+
+      case DartCObjectType.DartTypedData:
+        return _typedDataIntoDart(
+          value.as_typed_data.ty,
+          value.as_typed_data.values,
+          value.as_typed_data.length,
+          copy: true,
+        );
+
+      case DartCObjectType.DartExternalTypedData:
+        final externalTypedData = _typedDataIntoDart(
+          value.as_external_typed_data.ty,
+          value.as_external_typed_data.data,
+          value.as_external_typed_data.length,
+          copy: false,
+        );
+        _externalTypedDataFinalizer.attach(
+            externalTypedData, value.as_external_typed_data);
+        return externalTypedData;
+
+      case DartCObjectType.DartSendPort:
+      case DartCObjectType.DartCapability:
+      case DartCObjectType.DartNativePointer:
+      case DartCObjectType.DartUnsupported:
+      case DartCObjectType.DartNumberOfTypes:
+      default:
+        throw "Can't read invalid data type $ty";
     }
-    return Uint8List.fromList(ptr.asTypedList(len));
   }
 
-  bool get isSuccess => success > 0;
+  static dynamic _typedDataIntoDart(
+    int ty,
+    ffi.Pointer<ffi.Uint8> typedValues,
+    int nValues, {
+    required bool copy,
+  }) {
+    switch (ty) {
+      case DartTypedDataType.ByteData:
+        final view = typedValues.cast<ffi.Uint8>().asTypedList(nValues);
+        final bytes = copy ? view : Uint8List.fromList(view);
```

Comment:
⚠️ looks like it is flipped? if copy should use fromList, but it is the converse

That's why I do not like this style :( too easy to make a mistake

Reply:
Well spotted! Okay you made your point :+1: 

I didn't see conditional values as unordered parameters of the same type, will now avoid them when possible (especially repeatedly)

Reply:
Just pushed latest changes, only this should be missing, I'm trying something.